# Data Profiling

---

## Imports

#### Python libraries

In [2]:
import pandas as pd
pd.options.display.max_columns = 50

import sys

#### Asistin configuration

In [3]:
sys.path.append("..")

#### Ancillary modules

In [4]:
## Reload modules
%load_ext autoreload
%autoreload 2

from src.utils.utils import (
    json_dump_dict,
    count_vars,
    count_obs,
    count_unique_obs,
    vars_classif_by_type,
    data_profiling_numeric,
    data_profiling_categ,
    data_profiling_date,
    data_profiling_loc
)

from src.utils.data_dict import (
    data_dict
)

---

## Reading data

In [4]:
df_c5 = pd.read_csv("../incidentes-viales-c5.csv")

In [5]:
df_c5

folio fecha_creacion hora_creacion dia_semana  \
0        GA/160123/05714     23/01/2016      22:35:04     Sábado   
1        AO/160123/05826     23/01/2016      22:50:49     Sábado   
2        C4/160124/02187     24/01/2016      09:40:11    Domingo   
3        C4/160124/05723     24/01/2016      22:40:57    Domingo   
4        C4/160124/01334     24/01/2016      04:25:15    Domingo   
...                  ...            ...           ...        ...   
1383133  GA/201022/03443     22/10/2020      16:17:05     Jueves   
1383134  GA/201022/03492     22/10/2020      16:29:13     Jueves   
1383135  C5/201022/02030     22/10/2020      12:00:26     Jueves   
1383136  C5/201022/02039     22/10/2020      12:01:25     Jueves   
1383137  C5/201022/03364     22/10/2020      16:05:47     Jueves   

                                             codigo_cierre fecha_cierre  \
0        (A) La unidad de atención a emergencias fue de...   24/01/2016   
1        (A) La unidad de atención a emergencias fue de...   24/01/2016   
2        (N) La unidad de atención a emergencias fue de...   24/01/2016   
3        (N) La unidad de atención a emergencias fue de...   24/01/2016   
4        (A) La unidad de atención a emergencias fue de...   24/01/2016   
...                                                    ...          ...   
1383133  (A) La unidad de atención a emergencias fue de...   22/10/2020   
1383134  (A) La unidad de atención a emergencias fue de...   22/10/2020   
1383135  (D) El incidente reportado se registró en dos ...   22/10/2020   
1383136  (D) El incidente reportado se registró en dos ...   22/10/2020   
1383137  (D) El incidente reportado se registró en dos ...   22/10/2020   

         año_cierre mes_cierre hora_cierre    delegacion_inicio  \
0              2016      Enero    00:21:08  VENUSTIANO CARRANZA   
1              2016      Enero    04:40:37           CUAJIMALPA   
2              2016      Enero    11:13:03              TLALPAN   
3              2016      Enero    22:55:51  MAGDALENA CONTRERAS   
4              2016      Enero    05:03:59       MIGUEL HIDALGO   
...             ...        ...         ...                  ...   
1383133        2020    Octubre    16:58:49    GUSTAVO A. MADERO   
1383134        2020    Octubre    18:54:22  VENUSTIANO CARRANZA   
1383135        2020    Octubre    12:02:30         AZCAPOTZALCO   
1383136        2020    Octubre    12:05:59  VENUSTIANO CARRANZA   
1383137        2020    Octubre    16:10:13           IZTAPALAPA   

                            incidente_c4    latitud   longitud  \
0        accidente-choque sin lesionados  19.422113 -99.084643   
1        accidente-choque con lesionados  19.358320 -99.297641   
2        accidente-choque sin lesionados  19.217070 -99.219070   
3        accidente-choque sin lesionados  19.320580 -99.241010   
4        accidente-choque sin lesionados  19.452900 -99.215870   
...                                  ...        ...        ...   
1383133            lesionado-atropellado  19.486310 -99.118120   
1383134  accidente-choque con lesionados  19.427274 -99.115814   
1383135  accidente-choque sin lesionados  19.467180 -99.175250   
1383136  accidente-choque con lesionados  19.427280 -99.096860   
1383137  accidente-choque sin lesionados  19.345780 -99.062980   

         clas_con_f_alarma      tipo_entrada    delegacion_cierre  \
0               EMERGENCIA  BOTÓN DE AUXILIO  VENUSTIANO CARRANZA   
1        URGENCIAS MEDICAS  BOTÓN DE AUXILIO           CUAJIMALPA   
2               EMERGENCIA   LLAMADA DEL 066              TLALPAN   
3               EMERGENCIA   LLAMADA DEL 066  MAGDALENA CONTRERAS   
4               EMERGENCIA   LLAMADA DEL 066       MIGUEL HIDALGO   
...                    ...               ...                  ...   
1383133  URGENCIAS MEDICAS  BOTÓN DE AUXILIO    GUSTAVO A. MADERO   
1383134  URGENCIAS MEDICAS  BOTÓN DE AUXILIO  VENUSTIANO CARRANZA   
1383135         EMERGENCIA   LLAMADA DEL 911         AZCAPOTZALCO   
1383136  URGE

---

## Data profiling

### Generic exploration

#### Number of variables in data

In [6]:
count_vars(df_c5)

Número de variables en los datos --> 18


#### Number of observations in data

In [7]:
count_obs(df_c5)

Número de observaciones en los datos --> 1383138


#### Number of unique observations in data per column

In [8]:
count_unique_obs(df_c5)

folio                1383138
fecha_creacion          2637
hora_creacion         105887
dia_semana                 7
codigo_cierre              5
fecha_cierre            2621
año_cierre                 7
mes_cierre                12
hora_cierre           106827
delegacion_inicio         16
incidente_c4              26
latitud                82501
longitud               78984
clas_con_f_alarma          4
tipo_entrada               9
delegacion_cierre         16
geopoint              134288
mes                       12
dtype: int64

#### Classifying variables according to data type

In [9]:
vars_classif_by_type()

categoric   date_and_time coordenate
0              folio  fecha_creacion    latitud
1      codigo_cierre   hora_creacion   longitud
2  delegacion_inicio      dia_semana   geopoint
3       incidente_c4    fecha_cierre          -
4  clas_con_f_alarma      año_cierre          -
5       tipo_entrada      mes_cierre          -
6  delegacion_cierre     hora_cierre          -
7                  -             mes          -

None


### Data profilling - numeric variables

El dataset analizado no contiene variables numéricas.

### Data profilling - categoric variables

In [10]:
cat_vars = [var for var in data_dict if data_dict[var]["data_type"] == "categoric"]

data_profiling_categ(df_c5, cat_vars)

*************************************
** Conteo y proporción de entradas **
*************************************
Variable categorica -> folio


Observaciones Proporción
Categoría                                
AO/160421/00122              1       0.0%
C5/180716/08485              1       0.0%
C5/180405/06900              1       0.0%
BJ/160804/01876              1       0.0%
C5/191110/09070              1       0.0%
...                        ...        ...
C5/190228/08243              1       0.0%
C5/190119/02669              1       0.0%
C5/180424/03814              1       0.0%
C5/191231/05107              1       0.0%
C5/181126/08245              1       0.0%

[1383138 rows x 2 columns]

None



Variable categorica -> codigo_cierre


Observaciones Proporción
Categoría                                                                   
(A) La unidad de atención a emergencias fue des...         600108      43.4%
(D) El incidente reportado se registró en dos o...         495631      35.8%
(N) La unidad de atención a emergencias fue des...         271677      19.6%
(F) El operador/a o despachador/a identifican, ...          10626       0.8%
(I) El incidente reportado es afirmativo y se a...           5096       0.4%

None



Variable categorica -> delegacion_inicio


Observaciones Proporción
Categoría                                    
IZTAPALAPA                  223687      16.2%
GUSTAVO A. MADERO           144620      10.5%
CUAUHTEMOC                  141391      10.2%
MIGUEL HIDALGO              115864       8.4%
COYOACAN                    110146       8.0%
BENITO JUAREZ               104322       7.5%
ALVARO OBREGON              102071       7.4%
VENUSTIANO CARRANZA          94393       6.8%
TLALPAN                      87339       6.3%
AZCAPOTZALCO                 60823       4.4%
IZTACALCO                    58567       4.2%
XOCHIMILCO                   52575       3.8%
TLAHUAC                      33420       2.4%
CUAJIMALPA                   26714       1.9%
MAGDALENA CONTRERAS          20062       1.5%
MILPA ALTA                    6986       0.5%

None



Variable categorica -> incidente_c4


Observaciones Proporción
Categoría                                                              
accidente-choque sin lesionados                       760506      55.0%
accidente-choque con lesionados                       315031      22.8%
lesionado-atropellado                                 193553      14.0%
accidente-motociclista                                 55646       4.0%
accidente-volcadura                                    31766       2.3%
lesionado-accidente automovilístico                     6054       0.4%
accidente-persona atrapada / desbarrancada              5607       0.4%
accidente-ciclista                                      2718       0.2%
accidente-otros                                         2465       0.2%
accidente-choque con prensados                          2334       0.2%
cadáver-atropellado                                     1975       0.1%
accidente-vehiculo atrapado                             1121       0.1%
cadáver-accidente automovilístico                       1083       0.1%
accidente-vehículo atrapado-varado                       852       0.1%
detención ciudadana-atropellado                          702       0.1%
sismo-choque sin lesionados                              540       0.0%
accidente-vehiculo desbarrancado                         490       0.0%
sismo-choque con lesionados                              372       0.0%
detención ciudadana-accidente automovilístico            153       0.0%
mi ciudad-calle-incidente de tránsito                     55       0.0%
accidente-ferroviario                                     55       0.0%
accidente-monopatín                                       31       0.0%
sismo-persona atropellada                                 23       0.0%
Detención ciudadana-accidente automovilístico              3       0.0%
sismo-choque con prensados                                 2       0.0%
mi ciudad-taxi-incidente de tránsito                       1       0.0%

None



Variable categorica -> clas_con_f_alarma


Observaciones Proporción
Categoría                                  
EMERGENCIA                799037      57.8%
URGENCIAS MEDICAS         572619      41.4%
FALSA ALARMA               10630       0.8%
DELITO                       852       0.1%

None



Variable categorica -> tipo_entrada


Observaciones Proporción
Categoría                                 
LLAMADA DEL 911          739339      53.5%
LLAMADA DEL 066          463432      33.5%
BOTÓN DE AUXILIO          78630       5.7%
RADIO                     77749       5.6%
LLAMADA APP911             9483       0.7%
ZELLO                      6451       0.5%
REDES                      4697       0.3%
CÁMARA                     3276       0.2%
APLICATIVOS                  81       0.0%

None



Variable categorica -> delegacion_cierre


Observaciones Proporción
Categoría                                    
IZTAPALAPA                  225202      16.3%
GUSTAVO A. MADERO           147582      10.7%
CUAUHTEMOC                  141781      10.3%
MIGUEL HIDALGO              113910       8.2%
COYOACAN                    107251       7.8%
ALVARO OBREGON              104073       7.5%
BENITO JUAREZ               103602       7.5%
TLALPAN                      92122       6.7%
VENUSTIANO CARRANZA          91661       6.6%
AZCAPOTZALCO                 59776       4.3%
IZTACALCO                    59183       4.3%
XOCHIMILCO                   50642       3.7%
TLAHUAC                      33148       2.4%
CUAJIMALPA                   26268       1.9%
MAGDALENA CONTRERAS          19703       1.4%
MILPA ALTA                    7094       0.5%

None



***************************************
** Impresión de resumen de variables **
***************************************


Info                              folio  \
Num_Registros                   1383138   
Num_de_categorias               1383138   
Moda                    AO/140101/01144   
Valores_faltantes                     0   
Top1               [AO/160421/00122, 1]   
Top2               [C5/180716/08485, 1]   
Top3               [C5/180405/06900, 1]   

Info                                                   codigo_cierre  \
Num_Registros                                                1383138   
Num_de_categorias                                                  5   
Moda               (A) La unidad de atención a emergencias fue de...   
Valores_faltantes                                                  0   
Top1               [(A) La unidad de atención a emergencias fue d...   
Top2               [(D) El incidente reportado se registró en dos...   
Top3               [(N) La unidad de atención a emergencias fue d...   

Info                         delegacion_inicio  \
Num_Registros                          1383138   
Num_de_categorias                           16   
Moda                                IZTAPALAPA   
Valores_faltantes                          158   
Top1                      [IZTAPALAPA, 223687]   
Top2               [GUSTAVO A. MADERO, 144620]   
Top3                      [CUAUHTEMOC, 141391]   

Info                                            incidente_c4  \
Num_Registros                                        1383138   
Num_de_categorias                                         26   
Moda                         accidente-choque sin lesionados   
Valores_faltantes                                          0   
Top1               [accidente-choque sin lesionados, 760506]   
Top2               [accidente-choque con lesionados, 315031]   
Top3                         [lesionado-atropellado, 193553]   

Info                         clas_con_f_alarma               tipo_entrada  \
Num_Registros                          1383138                    1383138   
Num_de_categorias                            4                          9   
Moda                                EMERGENCIA            LLAMADA DEL 911   
Valores_faltantes                            0                          0   
Top1                      [EMERGENCIA, 799037]  [LLAMADA DEL 911, 739339]   
Top2               [URGENCIAS MEDICAS, 572619]  [LLAMADA DEL 066, 463432]   
Top3                     [FALSA ALARMA, 10630]  [BOTÓN DE AUXILIO, 78630]   

Info                         delegacion_cierre  
Num_Registros                          1383138  
Num_de_categorias                           16  
Moda                                IZTAPALAPA  
Valores_faltantes                          140  
Top1                      [IZTAPALAPA, 225202]  
Top2               [GUSTAVO A. MADERO, 147582]  
Top3                      [CUAUHTEMOC, 141781]

None


**Con base en el profiling realizado, nos damos cuenta de los siguientes aspectos importantes acerca de los datos:**
- Observaciones principales
    1. Sí tenemos un folio único para cada una de las observaciones.
    2. El 20% de los envíos de asistencia registrados en la base (271,677) son inútiles debido a que la ambulancia llega al lugar pero no hay emergencia que atender.
    3. El 55% de los accidentes no requieren que se envíe una ambulancia de asistencia.
- Observaciones secundarias
    1. Solamente el 0.8% de las llamadas son identificadas como "Falsas" antes de enviar asistencia (información útil para definir *baseline*)

### Data profilling - date variables

Current processing assumptions:
1. All time entries which 3rd character is not ":" will be ignored. The first two characters of the rest will be taken as the correct hour.

In [11]:
dat_vars = [var for var in data_dict if 
            (data_dict[var]["data_type"] == "date_and_time")
            & ("_cierre" not in var)
           ]

df_dat = data_profiling_date(df_c5, dat_vars)

*********************************
** General description of data **
*********************************


fecha_creacion hora_creacion dia_semana      mes dia_creacion  \
dtype                object        object     object    int64       object   
notnull_v           1383138       1383138    1383138  1383138      1383138   
missing_v                 0             0          0        0            0   
count_unique           2637            25          7       12           31   

             mes_creacion año_creacion  
dtype              object       object  
notnull_v         1383138      1383138  
missing_v               0            0  
count_unique           12            8

None
---------------------------------------------------------------------------
---------------------------------------------------------------------------



****************************
** Top repeated variables **
****************************


fecha_creacion                    hora_creacion                      \
               value count part_notnull         value  count part_notnull   
top_1     14/02/2020  1163         0.08            19  92287         6.67   
top_2     26/10/2018  1028         0.07            20  91329         6.60   
top_3     08/12/2017  1011         0.07            18  87975         6.36   
top_4     30/11/2019  1011         0.07            21  87059         6.29   
top_5     21/04/2018  1009         0.07            16  82826         5.99   

      dia_semana                        mes                      dia_creacion  \
           value   count part_notnull value   count part_notnull        value   
top_1    Viernes  229681        16.61    10  133641         9.66           14   
top_2     Sábado  214134        15.48     8  121123         8.76           15   
top_3     Jueves  198147        14.33     9  120581         8.72           13   
top_4  Miércoles  193857        14.02     3  120335         8.70           28   
top_5     Martes  189669        13.71    12  117616         8.50           08   

                          mes_creacion                      año_creacion  \
       count part_notnull        value   count part_notnull        value   
top_1  47933         3.47           10  133635         9.66         2018   
top_2  47930         3.47           08  121131         8.76         2019   
top_3  47095         3.40           09  120572         8.72         2017   
top_4  46372         3.35           03  120240         8.69         2016   
top_5  46212         3.34           12  117497         8.49         2015   

                            
        count part_notnull  
top_1  246579        17.83  
top_2  229500        16.59  
top_3  226530        16.38  
top_4  191007        13.81  
top_5  184818        13.36

None
---------------------------------------------------------------------------
---------------------------------------------------------------------------



---

### Data profilling - location variables

In [12]:
loc_vars = [var for var in data_dict if (data_dict[var]["data_type"] == "coordenate")]
data_profiling_loc(df_c5, loc_vars)

*********************************
** General description of data **
*********************************


latitud longitud geopoint
dtype      float64  float64   object
notnull_v  1382695  1382703  1382692
missing_v      443      435      446

None
***************************
** Coordinates precision **
***************************


latitud_nd                       longitud_nd                     
           value    count part_notnull       value   count part_notnull
top_1        8.0  1003515        72.58         8.0  997835        72.17
top_2        5.0   173878        12.58         5.0  173603        12.56
top_3        6.0    95810         6.93         6.0   99496         7.20
top_4       15.0    77353         5.59        14.0   77952         5.64
top_5        4.0    25508         1.84         4.0   26896         1.95

None
---------------------------------------------------------------------------
---------------------------------------------------------------------------



---

## *Notes*

### Creating data dictionary

#### From scratch

In [ ]:
data_dict = {}
for col in df_c5.columns:
    data_dict[col] = {}
    data_dict[col]["relevant"] = True
    data_dict[col]["data_type"] = "-"

In [ ]:
json_dump_dict(data_dict)

In [ ]:
data_dict

---

#### Change existing

In [7]:
json_dump_dict(data_dict)

{
    "folio": {
        "relevant": false,
        "data_type": "categoric"
    },
    "fecha_creacion": {
        "relevant": true,
        "data_type": "date_and_time"
    },
    "hora_creacion": {
        "relevant": true,
        "data_type": "date_and_time"
    },
    "dia_semana": {
        "relevant": true,
        "data_type": "date_and_time"
    },
    "codigo_cierre": {
        "relevant": true,
        "data_type": "categoric"
    },
    "fecha_cierre": {
        "relevant": false,
        "data_type": "date_and_time"
    },
    "año_cierre": {
        "relevant": false,
        "data_type": "date_and_time"
    },
    "mes_cierre": {
        "relevant": false,
        "data_type": "date_and_time"
    },
    "hora_cierre": {
        "relevant": false,
        "data_type": "date_and_time"
    },
    "delegacion_inicio": {
        "relevant": false,
        "data_type": "categoric"
    },
    "incidente_c4": {
        "relevant": true,
        "data_type": "categoric"
    },
 

In [8]:
for feat in data_dict:
    data_dict[feat]["model_relevant"] = False

In [9]:
json_dump_dict(data_dict)

{
    "folio": {
        "relevant": false,
        "data_type": "categoric",
        "model_relevant": false
    },
    "fecha_creacion": {
        "relevant": true,
        "data_type": "date_and_time",
        "model_relevant": false
    },
    "hora_creacion": {
        "relevant": true,
        "data_type": "date_and_time",
        "model_relevant": false
    },
    "dia_semana": {
        "relevant": true,
        "data_type": "date_and_time",
        "model_relevant": false
    },
    "codigo_cierre": {
        "relevant": true,
        "data_type": "categoric",
        "model_relevant": false
    },
    "fecha_cierre": {
        "relevant": false,
        "data_type": "date_and_time",
        "model_relevant": false
    },
    "año_cierre": {
        "relevant": false,
        "data_type": "date_and_time",
        "model_relevant": false
    },
    "mes_cierre": {
        "relevant": false,
        "data_type": "date_and_time",
        "model_relevant": false
    },
    "hora_c

---
---